File: examples/opencog_ml_example.c
/**
 * OpenCog ML Inference Engine C Example
 *
 * This example demonstrates how to use the OpenCog ML C API for RWKV inference.
 * It shows basic inference operations and state management.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include "../opencog_ml.h"

void print_usage(const char* program_name) {
    printf("Usage: %s <model_path>\n", program_name);
    printf("\nThis example demonstrates the OpenCog ML inference engine C API.\n");
    printf("You need to provide a path to an RWKV model file in ggml format.\n");
    printf("\nTo get a model:\n");
    printf("1. Download an RWKV model from https://huggingface.co/BlinkDL\n");
    printf("2. Convert it using: python python/convert_pytorch_to_ggml.py model.pth model.bin FP16\n");
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        print_usage(argv[0]);
        return 1;
    }

    const char* model_path = argv[1];
    
    printf("OpenCog ML Inference Engine C API Demo\n");
    printf("======================================\n");
    
    // Enable debug mode
    opencog_ml_set_debug_mode(true);
    
    // Print system information
    const char* system_info = opencog_ml_get_system_info();
    printf("System Info: %s\n", system_info);
    
    // Configure the inference engine
    struct opencog_ml_engine_config config = {
        .model_path = model_path,
        .n_threads = 4,
        .n_gpu_layers = 0,
        .enable_state_caching = true,
        .temperature = 1.0f,
        .top_k = 50,
        .top_p = 0.9f
    };
    
    printf("\nInitializing engine with model: %s\n", model_path);
    
    // Initialize the inference engine
    struct opencog_ml_engine* engine = opencog_ml_init_engine(&config);
    if (!engine) {
        printf("Error: Failed to initialize OpenCog ML engine\n");
        return 1;
    }
    
    printf("Engine initialized successfully!\n");
    
    // Get model information
    size_t vocab_size = opencog_ml_get_vocab_size(engine);
    size_t embed_size = opencog_ml_get_embed_size(engine);
    size_t state_size = opencog_ml_get_state_size(engine);
    
    printf("Vocabulary size: %zu\n", vocab_size);
    printf("Embedding size: %zu\n", embed_size);
    printf("State size: %zu\n", state_size);
    
    // Create initial state
    printf("\nCreating initial state...\n");
    struct opencog_ml_tensor* initial_state = opencog_ml_create_initial_state(engine);
    if (!initial_state) {
        printf("Error: Failed to create initial state\n");
        opencog_ml_free_engine(engine);
        return 1;
    }
    
    printf("Initial state created with size: %zu\n", initial_state->size);
    
    // Demonstrate single token inference
    printf("\n--- Single Token Inference ---\n");
    uint32_t test_token = 42;
    
    struct opencog_ml_inference_result* result = opencog_ml_infer_token(
        engine, test_token, initial_state
    );
    
    if (result && result->error == OPENCOG_ML_ERROR_NONE) {
        printf("Input token: %u\n", test_token);
        printf("Predicted next token: %u\n", result->predicted_token);
        printf("Confidence: %.4f\n", result->confidence);
        printf("Logits tensor size: %zu\n", result->logits->size);
        printf("Output state tensor size: %zu\n", result->state->size);
    } else {
        printf("Single token inference failed\n");
        if (result) {
            printf("Error code: %d\n", result->error);
        }
    }
    
    // Demonstrate sequence inference
    printf("\n--- Sequence Inference ---\n");
    uint32_t test_sequence[] = {1, 2, 3, 42, 100};
    size_t sequence_length = sizeof(test_sequence) / sizeof(test_sequence[0]);
    
    struct opencog_ml_inference_result* seq_result = opencog_ml_infer_sequence(
        engine, test_sequence, sequence_length, initial_state
    );
    
    if (seq_result && seq_result->error == OPENCOG_ML_ERROR_NONE) {
        printf("Input sequence: [");
        for (size_t i = 0; i < sequence_length; i++) {
            printf("%u", test_sequence[i]);
            if (i < sequence_length - 1) printf(", ");
        }
        printf("]\n");
        printf("Predicted next token: %u\n", seq_result->predicted_token);
        printf("Confidence: %.4f\n", seq_result->confidence);
    } else {
        printf("Sequence inference failed\n");
        if (seq_result) {
            printf("Error code: %d\n", seq_result->error);
        }
    }
    
    // Demonstrate custom sampling
    printf("\n--- Custom Sampling Inference ---\n");
    struct opencog_ml_inference_result* custom_result = opencog_ml_infer_with_sampling(
        engine, test_token, initial_state, 0.8f, 20, 0.95f
    );
    
    if (custom_result && custom_result->error == OPENCOG_ML_ERROR_NONE) {
        printf("Custom sampling result - predicted token: %u, confidence: %.4f\n",
               custom_result->predicted_token, custom_result->confidence);
    } else {
        printf("Custom sampling inference failed\n");
    }
    
    // Demonstrate engine cloning
    printf("\n--- Engine Cloning ---\n");
    struct opencog_ml_engine* cloned_engine = opencog_ml_clone_engine(engine, 2);
    if (cloned_engine) {
        printf("Engine cloned successfully\n");
        
        // Test cloned engine
        struct opencog_ml_inference_result* clone_result = opencog_ml_infer_token(
            cloned_engine, test_token, NULL
        );
        
        if (clone_result && clone_result->error == OPENCOG_ML_ERROR_NONE) {
            printf("Cloned engine inference successful: predicted token %u\n", 
                   clone_result->predicted_token);
        }
        
        // Clean up clone result
        if (clone_result) {
            opencog_ml_free_inference_result(clone_result);
        }
        
        // Free cloned engine
        opencog_ml_free_engine(cloned_engine);
    } else {
        printf("Failed to clone engine\n");
    }
    
    // Demonstrate state management
    printf("\n--- State Management ---\n");
    struct opencog_ml_tensor* new_state = opencog_ml_create_initial_state(engine);
    if (new_state) {
        printf("Created new state tensor\n");
        
        // Reset state
        bool reset_success = opencog_ml_reset_state(engine, new_state);
        printf("State reset: %s\n", reset_success ? "Success" : "Failed");
        
        // Clean up
        opencog_ml_free_tensor(new_state);
    }
    
    // Test error handling
    printf("\n--- Error Handling ---\n");
    enum opencog_ml_error_flags last_error = opencog_ml_get_last_error(engine);
    printf("Last error code: %d\n", last_error);
    
    // Clean up resources
    printf("\nCleaning up resources...\n");
    
    if (result) {
        opencog_ml_free_inference_result(result);
    }
    
    if (seq_result) {
        opencog_ml_free_inference_result(seq_result);
    }
    
    if (custom_result) {
        opencog_ml_free_inference_result(custom_result);
    }
    
    if (initial_state) {
        opencog_ml_free_tensor(initial_state);
    }
    
    opencog_ml_free_engine(engine);
    
    printf("\n=== Demo completed successfully! ===\n");
    printf("The OpenCog ML inference engine is now ready for integration\n");
    printf("with OpenCog's cognitive architecture.\n");
    
    return 0;
}

File: examples/opencog_ml_example.py
#!/usr/bin/env python3
"""
Example demonstrating OpenCog ML Inference Engine for RWKV

This example shows how to use the OpenCog ML interface as a pure inference 
engine for RWKV models. It demonstrates:
- Engine initialization
- Single token inference  
- Sequence inference
- Text generation
- State management
"""

import sys
import os

# Add the parent directory to the path to import opencog_ml
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))

try:
    from opencog_ml import OpenCogMLEngine, EngineConfig, Tensor, set_debug_mode, get_system_info
    import numpy as np
except ImportError as e:
    print(f"Error importing OpenCog ML: {e}")
    print("Make sure the library is built and accessible.")
    sys.exit(1)


def demonstrate_basic_inference(model_path: str):
    """Demonstrate basic inference operations."""
    print("=== OpenCog ML Basic Inference Demo ===")
    
    # Enable debug mode for verbose output
    set_debug_mode(True)
    
    # Print system information
    print(f"System Info: {get_system_info()}")
    
    # Configure the inference engine
    config = EngineConfig(
        model_path=model_path,
        n_threads=4,
        n_gpu_layers=0,  # Use CPU for this demo
        enable_state_caching=True,
        temperature=1.0,
        top_k=50,
        top_p=0.9
    )
    
    print(f"Initializing engine with model: {model_path}")
    
    try:
        # Initialize the inference engine
        engine = OpenCogMLEngine(config)
        
        print(f"Engine initialized successfully!")
        print(f"Vocabulary size: {engine.vocab_size}")
        print(f"Embedding size: {engine.embed_size}")
        print(f"State size: {engine.state_size}")
        
        # Create initial state
        print("\nCreating initial state...")
        initial_state = engine.create_initial_state()
        print(f"Initial state shape: {initial_state.shape}")
        
        # Demonstrate single token inference
        print("\n--- Single Token Inference ---")
        test_token = 42  # Example token ID
        result = engine.infer_token(test_token, initial_state)
        
        if result.is_successful():
            print(f"Input token: {test_token}")
            print(f"Predicted next token: {result.predicted_token}")
            print(f"Confidence: {result.confidence:.4f}")
            print(f"Logits shape: {result.logits.shape}")
            print(f"Output state shape: {result.state.shape}")
        else:
            print(f"Single token inference failed: {result.get_error_message()}")
        
        # Demonstrate sequence inference
        print("\n--- Sequence Inference ---")
        test_sequence = [1, 2, 3, 42, 100]  # Example token sequence
        seq_result = engine.infer_sequence(test_sequence, initial_state)
        
        if seq_result.is_successful():
            print(f"Input sequence: {test_sequence}")
            print(f"Predicted next token: {seq_result.predicted_token}")
            print(f"Confidence: {seq_result.confidence:.4f}")
        else:
            print(f"Sequence inference failed: {seq_result.get_error_message()}")
        
        # Demonstrate text generation
        print("\n--- Text Generation ---")
        prompt = [1, 2, 3]  # Simple prompt
        generated_tokens, confidences = engine.generate(
            prompt, 
            max_length=10, 
            temperature=0.8
        )
        
        print(f"Prompt: {prompt}")
        print(f"Generated sequence: {generated_tokens}")
        print(f"Confidence scores: {[f'{c:.3f}' for c in confidences]}")
        
        # Demonstrate batch inference
        print("\n--- Batch Inference ---")
        batch_sequences = [
            [1, 2, 3],
            [4, 5, 6, 7],
            [8, 9]
        ]
        
        batch_results = engine.batch_infer(batch_sequences)
        
        for i, (seq, result) in enumerate(zip(batch_sequences, batch_results)):
            if result.is_successful():
                print(f"Sequence {i+1}: {seq} -> predicted token: {result.predicted_token} (confidence: {result.confidence:.3f})")
            else:
                print(f"Sequence {i+1} failed: {result.get_error_message()}")
        
        # Demonstrate engine cloning for multi-threading
        print("\n--- Engine Cloning ---")
        cloned_engine = engine.clone(n_threads=2)
        print(f"Cloned engine created with vocab size: {cloned_engine.vocab_size}")
        
        # Test cloned engine
        clone_result = cloned_engine.infer_token(test_token)
        if clone_result.is_successful():
            print(f"Cloned engine inference successful: predicted token {clone_result.predicted_token}")
        
        print("\n=== Demo completed successfully! ===")
        
    except Exception as e:
        print(f"Error during demo: {e}")
        return False
    
    return True


def demonstrate_tensor_operations():
    """Demonstrate tensor operations."""
    print("\n=== Tensor Operations Demo ===")
    
    # Create tensors from different data sources
    print("Creating tensors from different data sources...")
    
    # From numpy array
    np_data = np.random.randn(5, 3).astype(np.float32)
    tensor1 = Tensor(np_data)
    print(f"Tensor from numpy: shape={tensor1.shape}, type={tensor1.tensor_type}")
    
    # From Python list
    list_data = [[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]]
    tensor2 = Tensor(list_data)
    print(f"Tensor from list: shape={tensor2.shape}, type={tensor2.tensor_type}")
    
    # Convert back to numpy and list
    np_converted = tensor1.to_numpy()
    list_converted = tensor2.to_list()
    
    print(f"Converted back to numpy: shape={np_converted.shape}")
    print(f"Converted back to list: {list_converted}")
    
    print("=== Tensor Operations Demo completed ===")


def main():
    """Main function."""
    if len(sys.argv) != 2:
        print("Usage: python opencog_ml_example.py <model_path>")
        print("\nThis example demonstrates the OpenCog ML inference engine.")
        print("You need to provide a path to an RWKV model file in ggml format.")
        print("\nTo get a model:")
        print("1. Download an RWKV model from https://huggingface.co/BlinkDL")
        print("2. Convert it using: python python/convert_pytorch_to_ggml.py model.pth model.bin FP16")
        sys.exit(1)
    
    model_path = sys.argv[1]
    
    # Check if model file exists
    if not os.path.exists(model_path):
        print(f"Error: Model file not found: {model_path}")
        sys.exit(1)
    
    print("OpenCog ML Inference Engine Demo")
    print("=" * 50)
    
    # Demonstrate tensor operations (doesn't require a model)
    demonstrate_tensor_operations()
    
    # Demonstrate inference operations (requires a model)
    print(f"\nLoading model from: {model_path}")
    success = demonstrate_basic_inference(model_path)
    
    if success:
        print("\n🎉 All demonstrations completed successfully!")
        print("\nThe OpenCog ML inference engine is now ready for integration")
        print("with OpenCog's cognitive architecture.")
    else:
        print("\n❌ Some demonstrations failed. Check the error messages above.")
        sys.exit(1)


if __name__ == "__main__":
    main()